### PR TITLE
 qemu.tests.qemu_guest_agent:add guest agent read/write file case

### PR DESCRIPTION
**Pull request by xiangchunfu**
This patch depend on another patch 
1.open a file via agent command guest-file-open
2.write content to file via agent command guest-file-write
3.read content from file via agent command guest-file-read
4.close current file via agent command guest-file-close

id:1468931 1468933 1468932 1468524

Signed-off-by: Xiangchun Fu <xfu@redhat.com>

### CODE REVIEW

**Comment by yanan-fu** [comment]
@xiangchunfu  please do not add depends on in commit info.

**Comment by yanan-fu** [comment]
Please remove  the second  commit and update the first one.

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -777,6 +777,50 @@ def gagent_check_fsfreeze(self, test, params, env):
...
+        ret_write = self.gagent.writefile(write_args)
+        close_args_w = {"handle": ret}
+        self.gagent.closefile(close_args_w)
+
+        #mode r+ open an exist file only
+        open_args_r = {"path": agent_file, "mode": "r+"}
+        ret = int(self.gagent.openfile(open_args_r))
+        read_args = {"handle": ret, "count": 1024}
+        ret_read = self.gagent.readfile(read_args)
+        if not ret_read["buf-b64"] == write_str:
```
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py]: if ret_read["buf-b64"] != write_str  ?

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -777,6 +777,50 @@ def gagent_check_fsfreeze(self, test, params, env):
...
+            test.fail("Read the contents and files do not match")
+        close_args_r = {"handle": ret}
+        self.gagent.closefile(close_args_r)
```
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py]: Checked the related function in https://github.com/avocado-framework/avocado-vt/pull/1116, use  success_resp=False as default.  I am puzzled if there has exception when open/close/read/write file ?

Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: @yanan-fu  Thanks,  I have updated this part in patch 1116

File: qemu/tests/qemu_guest_agent.py (hunk 1 above)
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: if not ret_read["buf-b64"] == write_str:
    test.fail("Read the contents and files do not match")

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -777,6 +777,50 @@ def gagent_check_fsfreeze(self, test, params, env):
...
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environmen.
+        """
+        if not self.env or not self.params:
+            test.error("You should run 'setup' method before test")
+
+        if not (self.vm and self.vm.is_alive()):
+            vm = self.env.get_vm(self.params["main_vm"])
+            vm.verify_alive()
+            self.vm = vm
```
[the remark this answers is not in the document]
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: As guest-agent command need to connect agent service in guest.   To make sure guest works. so add this checking.

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -777,6 +777,50 @@ def gagent_check_fsfreeze(self, test, params, env):
...
+        4) open file with mode "r+" in guest.
+        5) read file
+        6) close file
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environmen.
+        """
+        if not self.env or not self.params:
+            test.error("You should run 'setup' method before test")
```
[the remark this answers is not in the document]
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: @yanan-fu  Ensure test environment is available.  It seems that not all cases are needed.   Should I remove it?

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -775,6 +775,50 @@ def gagent_check_fsfreeze(self, test, params, env):
...
+        open_args_w = {"path": agent_file, "mode": "w+"}
+        ret = int(self.gagent.openfile(open_args_w))
+        write_str = "aGVsbG8gd29ybGQhCg=="
+        write_args = {"handle": ret, "buf-b64": write_str}
+        ret_write = self.gagent.writefile(write_args)
+        close_args_w = {"handle": ret}
+        self.gagent.closefile(close_args_w)
+
+        #mode r+ open an exist file only
+        open_args_r = {"path": agent_file, "mode": "r+"}
```
Comment (dinghuiqing) [inline qemu/tests/qemu_guest_agent.py]: Polarion case use "mode":"r" to open the file.  Suggest to define the mode in cfg and do not use hard code.

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -775,6 +775,50 @@ def gagent_check_fsfreeze(self, test, params, env):
...
+        ret = int(self.gagent.openfile(open_args_r))
+        read_args = {"handle": ret, "count": 1024}
+        ret_read = self.gagent.readfile(read_args)
+        if not ret_read["buf-b64"] == write_str:
+            test.fail("Read the contents and files do not match")
+        close_args_r = {"handle": ret}
+        self.gagent.closefile(close_args_r)
```
Comment (dinghuiqing) [inline qemu/tests/qemu_guest_agent.py]: At end of the case, the guest file should be deleted and recover the env to the original.

Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: @dinghuiqing  Thanks.   I have updated this part of code.

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: @dinghuiqing  Thanks,  Before I and xiywang discussed this issue.  she think this  isn't hard code.  maybe other .py file will use this mode as well.  I will reconsider  put this mode to cfg.

**Comment by xiangchunfu** [comment]
Already updated this patch again.  please review it. Thanks!

**Comment by xiangchunfu** [comment]
(1/2) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.qemu_guest_agent.virtio_serial.check_guest_file: PASS (418.97 s)
 (2/2) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.qemu_guest_agent.isa_serial.check_guest_file: PASS (421.96 s)

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: @yanan-fu  For this case, don't need to check env and params. If env is unavailable. framework will output error message to log.  so remove this part code.

File: qemu/tests/qemu_guest_agent.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: @yanan-fu  the same process as above.  Could you help review it again?

**Comment by xiangchunfu** [comment]
Xiagao will re-submit related patch.  In order to avoid conflicts, I will close this patch.

**Pull request closed without merging**